ably/realtime_channel.go

```diff
@@ -2,130 +2,342 @@ package ably
+	return c.PublishAll([]*proto.Message{{Name: name, Data: data}}, listen)
+}
+
+// PublishAll publishes all given messages on the channel at once.
+// PublishAll does not block.
+//
+// If listen channel is non-nil it will be used to notify whether the messages
+// were published succesfully with nil error value; when the operation failed,
+// the error that caused the failure will be sent on the listen channel.
+//
+// This implicitly attaches the channel if it's not already attached.
+func (c *RealtimeChannel) PublishAll(messages []*proto.Message, listen chan<- error) error {
+	if err := c.Attach(); err != nil {
+		return err
+	}
 	msg := &proto.ProtocolMessage{
-		Action:  proto.ActionMessage,
-		Channel: c.Name,
-		Messages: []*proto.Message{
-			{Name: name, Data: data},
-		},
+		Action:   proto.ActionMessage,
+		Channel:  c.Name,
+		Messages: messages,
 	}
-	return c.client.send(msg)
+	switch c.State() {
+	case StateChanInitialized, StateChanAttaching:
+		if c.client.opts.NoQueueing {
+			return errQueueing
+		}
+		c.queue.Enqueue(msg, listen)
+		return nil
+	case StateChanAttached:
+	default:
+		return &Error{Code: 90001}
+	}
+	return c.client.Connection.send(msg, listen)
+}
+
+// State gives current state of the channel.
+func (c *RealtimeChannel) State() int {
+	c.state.Lock()
+	state := c.state.current
+	c.state.Unlock()
+	return state
+}
+
+// Reason gives the last error that caused channel transition to failed state.
+func (c *RealtimeChannel) Reason() error {
+	c.state.Lock()
+	err := c.state.err
+	c.state.Unlock()
+	return err
 }
 
 func (c *RealtimeChannel) notify(msg *proto.ProtocolMessage) {
 	switch msg.Action {
 	case proto.ActionAttached:
-		c.setState(ChanStateAttached)
+		c.state.syncSet(StateChanAttached, nil)
+		c.queue.Flush()
 	case proto.ActionDetached:
-		c.setState(ChanStateDetached)
+		c.state.syncSet(StateChanDetached, nil)
 	case proto.ActionPresence:
-		// TODO
+		Log.Printf(LogError, "TODO: implement RealtimePresence: %v", msg)
 	case proto.ActionError:
-		c.setState(ChanStateFailed)
-		c.Err = msg.Error
-		// TODO c.Close()
+		c.state.syncSet(StateChanFailed, newErrorProto(msg.Error))
+		c.queue.Fail(newErrorProto(msg.Error))
+		// TODO recovery
 	case proto.ActionMessage:
-		c.listenMtx.RLock()
-		defer c.listenMtx.RUnlock()
-		for _, m := range msg.Messages {
-			if l, ok := c.listeners[""]; ok {
+		c.listenMtx.Lock()
+		for _, msg := range msg.Messages {
+			if l, ok := c.listen[""]; ok {
 				for ch := range l {
-					ch <- m
+					select {
+					case ch <- msg:
+					default:
+						Log.Printf(LogWarn, "dropped %q message due to slow receiver", msg.Name)
+					}
 				}
 			}
-			if l, ok := c.listeners[m.Name]; ok {
+			if l, ok := c.listen[msg.Name]; ok {
 				for ch := range l {
-					ch <- m
+					select {
+					case ch <- msg:
+					default:
+						Log.Printf(LogWarn, "dropped %q message due to slow receiver", msg.Name)
+					}
 				}
 			}
```

Comment:
> I don’t want to get rid of slow receivers, I want slow receivers not to leak goroutines started by us

Why would the user being slow lead to them leaking goroutines? The proposed `Subscription` would likely have a `Close()` method which would stop the goroutine, and the user should be encouraged to defer that call, this is a pretty standard pattern:

```go
sub, err := c.Subscribe("foo")
if err != nil {
  return err
}
defer sub.Close()
```

> I consider slow receivers as programming error (insufficient buffering, too much blocking operations) on the user side which I would not want to handle.

You seem to want the user to do extra work when it is not necessary. I'd say it is quite typical for users to want to receive events and then take a non-trivial amount of time to handle those events.

```go
sub, _ := c.Subscribe("foo")
defer sub.Close()
for event := range sub.Events {
  // do some non-trivial processing
}
```

You are expecting the user to do one of two things:

1. Use a buffered channel. How big should this buffer be? I need to predict the rate of incoming events and the time it takes to process an event to determine an appropriate buffer size, and I probably don't know either of those at compile time. Too small a buffer and I start losing messages, which may have serious consequences.

2. Use an unbuffered channel and not be a slow receiver. So if I am processing events in non-trivial time I am going to need to create a goroutine per event. And if I also want to process them in order, I am going to need my own queueing goroutine to avoid blocking the channel, which is more work for me.

I think if you are still against supporting slow receivers, I think you need to make the above considerations clear to the user.